### PR TITLE
Update to 0.0.12

### PR DESCRIPTION
**Pull request by zqevans**
- Add LM training and inference
- Demos for generative models are now peak-normalized to avoid clipping
- Add pretransforms for pre-trained Audiocraft compression models like Encodec
- Add more features/properties for discrete autoencoders
- Add activation checkpointing to Oobleck and DiT
- Improved autoencoder chunking logic
- add AdaLN global conditioning to transformer/DiT
- 

- Removed Audiocraft import from autoencoders to improve loading time

### CODE REVIEW

File: stable_audio_tools/interface/gradio.py [hunk 1]
```diff
@@ -183,7 +187,7 @@ def progress_callback(callback_info):
 
     # Convert to WAV file
     audio = rearrange(audio, "b d n -> d (b n)")
-    audio = audio.clamp(-1, 1).mul(32767).to(torch.int16).cpu()
+    audio = audio.to(torch.float32).div(torch.max(torch.abs(audio))).clamp(-1, 1).mul(32767).to(torch.int16).cpu()
```
Comment (julian-parker) [inline stable_audio_tools/interface/gradio.py:190]: We could probably put this transformation into a utility function

File: stable_audio_tools/models/autoencoders.py [hunk 2]
```diff
@@ -13,51 +12,16 @@
...
-from .bottleneck import Bottleneck
+from .blocks import SnakeBeta
+from .bottleneck import Bottleneck, DiscreteBottleneck
 from .diffusion import ConditionedDiffusionModel, DAU1DCondWrapper, UNet1DCondWrapper, DiTWrapper
 from .factory import create_pretransform_from_config, create_bottleneck_from_config
 from .pretransforms import Pretransform, AutoencoderPretransform
 
-def snake_beta(x, alpha, beta):
-    return x + (1.0 / (beta + 0.000000001)) * pow(sin(x * alpha), 2)
+def checkpoint(function, *args, **kwargs):
```
Comment (julian-parker) [inline stable_audio_tools/models/autoencoders.py:21]: This is left from my patched autoencoder commit I think

Comment (julian-parker) [reply stable_audio_tools/models/autoencoders.py:21]: Never mind, I see its used elsewhere

File: stable_audio_tools/training/musicgen.py [hunk 3]
```diff
@@ -203,16 +203,17 @@ def on_train_batch_end(self, trainer, module: MusicGenTrainingWrapper, outputs,
...
+                    fakes = module.musicgen_model.generate(prompts, progress=True)
 
                 # Put the demos together
                 fakes = rearrange(fakes, 'b d n -> d (b n)')
 
                 log_dict = {}
                 
                 filename = f'demo_cfg_{cfg_scale}_{trainer.global_step:08}.wav'
-                fakes = fakes.clamp(-1, 1).mul(32767).to(torch.int16).cpu()
+                fakes = fakes.clamp(-1, 1).mul(32766).to(torch.int16).cpu()
```
Comment (zqevans) [inline stable_audio_tools/training/musicgen.py:216]: This should be replaced with a cast to torch.float32 before the hard-clip

File: stable_audio_tools/models/dit.py [hunk 4]
```diff
@@ -187,30 +179,38 @@ def _forward(
...
-            output = self.transformer(x, prepend_cond=prepend_inputs, cross_attn_cond=cross_attn_cond, cross_attn_cond_mask=cross_attn_cond_mask, mask=mask, **kwargs)
-        elif self.transformer_type == "x-transformers" or self.transformer_type == "continuous_transformer":
-            output = self.transformer(x, prepend_embeds=prepend_inputs, context=cross_attn_cond, context_mask=cross_attn_cond_mask, mask=mask, prepend_mask=prepend_mask, **kwargs)
+        extra_args = {}
+
+        if self.global_cond_type == "adaLN":
+            extra_args["global_cond"] = global_embed
+
+        if self.patch_size > 1:
+            x = rearrange(x, "b (t p) c -> b t (p c)", p=self.patch_size)
```
Comment (julian-parker) [inline stable_audio_tools/models/dit.py:204]: Just checking - did you test that this patching transform does what's intended and perfectly cancels with the one on L.212? I just did a quick test and it seems like they might not:
```
import torch
from einops import rearrange

batch = 5
channels = 3
length = 24
patch_size = 4

x = torch.arange(0,length).unsqueeze(0).unsqueeze(-1)
x = x.expand(batch, -1, channels).clone()
x[:,:,0] = -1 * x[:,:,0] 
x[:,:,2] = 100

ref = x.clone()

x = rearrange(x, "b (t p) c -> b t (p c)", p=patch_size)
x = rearrange(x, "b t c -> b c t")
x = rearrange(x, "b (c p) t -> b c (t p)", p=patch_size)
x = rearrange(x, "b c t-> b t c") # needed to compare to original (we conv in the real code)

assert torch.all(ref == x)
```

File: stable_audio_tools/training/diffusion.py [hunk 5]
```diff
@@ -546,7 +549,7 @@ def on_train_batch_end(self, trainer, module: DiffusionCondTrainingWrapper, outp
                 audio_inputs = rearrange(audio_inputs, 'b d n -> d (b n)')
 
                 filename = f'demo_audio_cond_{trainer.global_step:08}.wav'
-                audio_inputs = audio_inputs.clamp(-1, 1).mul(32767).to(torch.int16).cpu()
+                audio_inputs = audio_inputs.div(torch.max(torch.abs(audio_inputs))).mul(32767).to(torch.int16).cpu()
```
Comment (julian-parker) [inline stable_audio_tools/training/diffusion.py:552]: float32

File: stable_audio_tools/training/diffusion.py [hunk 6]
```diff
@@ -833,7 +836,7 @@ def on_train_batch_end(self, trainer, module: DiffusionCondTrainingWrapper, outp
                 log_dict = {}
                 
                 filename = f'demo_cfg_{cfg_scale}_{trainer.global_step:08}.wav'
-                fakes = fakes.clamp(-1, 1).mul(32767).to(torch.int16).cpu()
+                fakes = fakes.div(torch.max(torch.abs(fakes))).mul(32767).to(torch.int16).cpu()
```
Comment (julian-parker) [inline stable_audio_tools/training/diffusion.py:839]: float32

File: stable_audio_tools/training/diffusion.py [hunk 7]
```diff
@@ -1060,7 +1067,7 @@ def on_train_batch_end(self, trainer, module: DiffusionAutoencoderTrainingWrappe
         log_dict = {}
         
         filename = f'recon_{trainer.global_step:08}.wav'
-        reals_fakes = reals_fakes.clamp(-1, 1).mul(32767).to(torch.int16).cpu()
+        reals_fakes = reals_fakes.div(torch.max(torch.abs(reals_fakes))).mul(32767).to(torch.int16).cpu()
```
Comment (julian-parker) [inline stable_audio_tools/training/diffusion.py:1070]: float32

File: stable_audio_tools/training/diffusion.py [hunk 8]
```diff
@@ -1077,7 +1084,7 @@ def on_train_batch_end(self, trainer, module: DiffusionAutoencoderTrainingWrappe
                 initial_latents = module.diffae_ema.ema_model.pretransform.encode(encoder_input)
                 first_stage_fakes = module.diffae_ema.ema_model.pretransform.decode(initial_latents)
                 first_stage_fakes = rearrange(first_stage_fakes, 'b d n -> d (b n)')
-                first_stage_fakes = first_stage_fakes.clamp(-1, 1).mul(32767).to(torch.int16).cpu()
+                first_stage_fakes = first_stage_fakes.div(torch.max(torch.abs(first_stage_fakes))).to(torch.int16).cpu()
```
Comment (julian-parker) [inline stable_audio_tools/training/diffusion.py:1087]: wrong transform?

File: stable_audio_tools/training/diffusion.py [hunk 9]
```diff
@@ -1400,7 +1407,7 @@ def on_train_batch_end(self, trainer, module: DiffusionAutoencoderTrainingWrappe
         log_dict = {}
         
         filename = f'recon_{trainer.global_step:08}.wav'
-        reals_fakes = reals_fakes.clamp(-1, 1).mul(32767).to(torch.int16).cpu()
+        reals_fakes = reals_fakes.div(torch.max(torch.abs(reals_fakes))).mul(32767).to(torch.int16).cpu()
```
Comment (julian-parker) [inline stable_audio_tools/training/diffusion.py:1410]: float32

File: stable_audio_tools/training/musicgen.py (hunk 3 above)
Comment (julian-parker) [inline stable_audio_tools/training/musicgen.py:216]: float32

**Review by julian-parker — APPROVED** [review]

File: stable_audio_tools/models/dit.py (hunk 4 above)
Comment (zqevans) [reply stable_audio_tools/models/dit.py:204]: aha, the `p` and `c` should be swapped in the second half of this.

File: stable_audio_tools/interface/gradio.py (hunk 1 above)
Comment (zqevans) [reply stable_audio_tools/interface/gradio.py:190]: yeah, I'd rather do that in the dev repo and bring it over later
